src/lib/yang/alarms.lua

```diff
@@ -101,26 +140,207 @@ function declare_alarm (alarm)
    return alarm
 end
 
----
+-- Raise alarm.
+
+-- The entry with latest time-stamp in this list MUST correspond to the leafs
+-- 'is-cleared', 'perceived-severity' and 'alarm-text' for the alarm.
+-- The time-stamp for that entry MUST be equal to the 'last-changed' leaf.
+local function add_status_change (alarm, status)
+   alarm.status_change = alarm.status_change or {}
+   alarm.perceived_severity = status.perceived_severity
+   alarm.alarm_text = status.alarm_text
+   alarm.last_changed = status.time
+   state.alarm_list.last_changed = status.time
+   table.insert(alarm.status_change, status)
+end
+
+-- Creates a new alarm.
+--
+-- The alarm is retrieved from the db of predefined alarms. Default values got
+-- overridden by args. Additional fields are initialized too and an initial
+-- status change is added to the alarm.
+local function new_alarm (key, args)
+   local ret = assert(alarm_list:retrieve(key, args), 'Not supported alarm')
+   local status = {
+      time = iso_8601(),
+      perceived_severity = args.perceived_severity or ret.perceived_severity,
+      alarm_text = args.alarm_text or ret.alarm_text,
+   }
+   add_status_change(ret, status)
+   ret.last_changed = assert(status.time)
+   ret.time_created = assert(ret.last_changed)
+   ret.is_cleared = args.is_cleared
+   ret.operator_state_change = {}
+   state.alarm_list.number_of_alarms = state.alarm_list.number_of_alarms + 1
+   return ret
+end
+
+-- Adds alarm to state.alarm_list.
+local function create_alarm (key, args)
+   local alarm = assert(new_alarm(key, args))
+   state.alarm_list.alarm[key] = alarm
+end
+
+-- The following state changes creates a new status change:
+--   - changed severity (warning, minor, major, critical).
+--   - clearance status, this also updates the 'is-cleared' leaf.
+--   - alarm text update.
+local function needs_status_change (alarm, args)
+   if alarm.is_cleared ~= args.is_cleared then
+      return true
+   elseif args.perceived_severity and
+          alarm.perceived_severity ~= args.perceived_severity then
+      return true
+   elseif args.alarm_text and alarm.alarm_text ~= args.alarm_text then
+      return true
+   end
+   return false
+end
+
+-- An alarm gets updated if it needs a status change.  A status change implies
+-- to add a new status change to the alarm and update the alarm 'is_cleared'
+-- flag.
+local function update_alarm (alarm, args)
+   if needs_status_change(alarm, args) then
+      local status = {
+         time = assert(iso_8601()),
+         perceived_severity = assert(args.perceived_severity or alarm.perceived_severity),
+         alarm_text = assert(args.alarm_text or alarm.alarm_text),
+      }
+      add_status_change(alarm, status)
+      alarm.is_cleared = args.is_cleared
+   end
+end
+
+-- Check up if the alarm already exists in state.alarm_list.
+local function lookup_alarm (key)
+   return state.alarm_list.alarm[key]
+end
+
+-- Codec serializes nil values as empty strings. Convert them back to nil values.
```

Comment:
Why does the codec do this?  You can't detect the presence or absence of a nil-valued property anyway.

Reply:
Some arguments in `alarms_codec:raise` might be nil. For instance:

```
alarms:raise(key, {perceived_severity='minor'})
```

In this call `alarm_text` is nil.

`codec:string` cannot encode a `nil` value, so I have to encode nil arguments as something (an empty string). If I wouldn't do it and let `finish()` write only set values, I could not distinguish which argument is which.

Since all arguments have a value (even when they were originally `nil`), I do this conversion back to nil. If I wouldn't do it, in the example above, `alarm_text` will be empty and the description text of the alarm will get overwritten to an empty string.

Reply:
Could you make the codec define a `maybe_string` type?  Like if it writes UINT32_MAX for the length, then it's nil, otherwise it's a string.  dunno.  E.g.

```lua
local UINT32_MAX = 0xffffffff
function encoder:maybe_string(str)
   if str == nil then
      self:uint32(UINT32_MAX)
   else
      self:string(str)
   end
end

...

   function decoder:maybe_string(str)
      local len = self:uint32()
      if len == UINT32_MAX then return nil end
      return ffi.string(self:read(len), len)
   end
```

then use maybe_string wherever you would want to encode/decode a value that may be a string or nil.  That way you can actually encode the empty string if you need to!